### PR TITLE
fix(core): keep cursive words joined across a run face change

### PR DESCRIPTION
**Pull request by jan-kubica**
A cursive word split mid-word by a run boundary renders as isolated letter
forms whenever the two sides resolve to a different font face. Word joins
straight through the same boundary. This fixes that, and adds the invariant
that should have caught it.

## The defect

Browsers shape across an inline box boundary only while no shaping-relevant
property changes. Measured in Chrome against the exact span structure the
painter emits:

| boundary between two runs | painted width | verdict |
| --- | --- | --- |
| whole word, one run | 58.38px | baseline |
| identical styling | 58.38px | shapes across |
| colour + underline | 58.38px | shapes across |
| nested `<a>` (hyperlink) | 58.38px | shapes across |
| **bold** | 68.34px | joining broken |
| **italic** | 61.12px | joining broken |
| **font-size** | 68.62px | joining broken |
| **font-family** | 69.53px | joining broken |

So tracked-change and comment marks were never affected: those differ in colour
and decoration, not face. Bold on a letter inside a word is the everyday case
that breaks, and it cannot be fixed by shaping alone because the two halves
genuinely need different faces.

The repair is a zero-width joiner on each side, each in its own span wearing
that side's styles, which measured identically to the joined baseline.

## Why the joiner is not in the run's text

`data-pm-start` spans map DOM text offsets back to document positions for
hit-testing. Appending a character to a run's text node would desync every
offset after it, and compensating at each mapping site is the kind of
hand-maintained mirror that drifts. A separate span carries no pm attributes,
so the offset contract is untouched. Pinned by a test.

## Joining data is generated, not written

`ArabicShaping.txt` lists 4 Transparent characters. The other 375 ranges'
worth inherit it from General_Category Mn/Me/Cf in a different file, so a
hand-authored table would have misclassified every Arabic harakat, and the
failure would have shown up as mis-shaped text rather than a red test. Both UCD
sources are pinned with checksums; `generate:joining-types:check` fails if the
committed table drifts.

Face equality is decided by running `applyRunStyles` over a probe element and
diffing the result, rather than re-deriving the face from run fields, so there
is no second copy of the font decisions. An unrecognised property counts as a
face change: that adds a harmless joiner instead of silently skipping a repair.

## The invariant

Folio measures text with canvas and paints it as DOM text the browser shapes
itself. Two engines, and nothing compared their outputs, so they could disagree
while the whole suite stayed green.

`tests/visual/measure-parity.spec.ts` stamps the measurer's width on each line
and compares it against a Range over the line's content. Both numbers come from
one layout pass in the same browser, so unlike a screenshot baseline it cannot
go flaky on cross-machine font rendering.

It earned its place immediately, catching a real bug in this branch:
`Object.entries()` on a `CSSStyleDeclaration` yields indexed entries, so in a
browser the face comparison was diffing the list of property names rather than
their values, and every colour-only boundary looked like a face change. Unit
tests could not see it because the test fake's `style` is a plain object. Fixed,
with a regression test that drives a real-shaped declaration.

Counts are pinned rather than asserted non-zero: the fixture must produce
exactly 6 joiner spans, so both a false positive and a repair that stops
happening fail the build.

## Known residual

The measurer cannot see this repair. Canvas `measureText` ignores a joiner at a
string edge (measuring the halves with and without one returns bit-identical
widths) while DOM layout applies the joined forms, so a repaired word paints
narrower than was reserved. The error is bounded, arises only on words
containing a face change, and is conservative: layout over-reserves, so a line
may break marginally early and can never overflow. The spec pins that direction
so it cannot silently invert. Closing it needs widths from a real shaper rather
than canvas, which is separate work.

## Verification

- `bun run test`: docx-core 53, core 4521, react 156, vue 22, agents 98,
  scripts 100. Zero failures.
- `bunx playwright test --project=measure-parity`: 3 passed.
- `bun --filter @stll/folio-core typecheck`, `oxlint`, and
  `generate:joining-types:check` all clean.

## Follow-ups not in this PR

- Add `generate:joining-types:check` and the `measure-parity` project to CI so
  both actually gate.
- A fresh checkout needs `bun run specifications:fetch` before regenerating the
  table, same as the other pinned sources.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved rendering of cursive scripts, including Arabic, Syriac, N’Ko and Adlam, when text changes style within a word.
  - Preserved more accurate letter connections across formatting, colour and text-run boundaries.
  - Improved consistency between measured line widths and browser-rendered content.

- **Tests**
  - Added broader coverage for cursive text rendering, boundary cases and visual measurement accuracy.

### CODE REVIEW

File: packages/core/src/layout-painter/renderParagraph.ts [hunk 1]
```diff
@@ -2173,7 +2200,7 @@ export function renderLine(
             break;
           }
           if (isTextRun(next)) {
-            lineEl.append(renderLineTextRun(next));
+            lineEl.append(...withJoiners(renderLineTextRun(next), next));
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/core/src/layout-painter/renderParagraph.ts:2212]: Fixed. Confirmed the mechanism: a flex item is its own formatting context, so shaping cannot cross it and a joiner span there would join nothing while still adding a node. `withJoiners` now returns the run element unchanged once the line has been promoted to flex.

Worth recording that this is narrower than a fix. Cursive words already come apart on flex-promoted lines regardless of this PR, because the runs themselves are separate flex items. Keeping a run and its joiners together in one inline item would not help either, since the break is between two runs rather than between a run and its joiner. Repairing it means the tab layout not making each run a flex item, which is out of scope here.

File: scripts/generate-joining-types.ts [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+const readCachedSource = async (entry: SourceEntry): Promise<string> => {
+  const absolute = path.join(REPO_ROOT, entry.cachePath);
+  try {
+    return await readFile(absolute, "utf8");
+  } catch {
+    throw new GenerateJoiningTypesError({
+      message: `Cached source missing for \`${entry.id}\` at ${entry.cachePath}. Run \`bun run specifications:fetch\` first.`,
+    });
+  }
+};
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply scripts/generate-joining-types.ts:116]: Fixed. `SourceEntry` now carries `sha256`, and `readCachedSource` verifies the file against the manifest digest before parsing, naming the file to delete when it mismatches.

Verified the guard fires rather than assuming it: appending one byte to the cached `ArabicShaping.txt` makes `check` fail with the expected-versus-read digests, and it passes again after a re-fetch.

File: tests/visual/measure-parity.spec.ts [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        const textNodes: Text[] = [];
+        let node = walker.nextNode();
+        while (node !== null) {
+          textNodes.push(node as Text);
+          node = walker.nextNode();
+        }
+        for (let j = textNodes.length - 1; j >= 0; j--) {
+          // SAFETY: j is inside the array bounds.
+          const textNode = textNodes[j]!;
+          const trimmed = (textNode.data ?? "").replace(/[\s​]+$/u, "");
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply tests/visual/measure-parity.spec.ts]: Fixed — the class is now `[\s​]`.

File: tests/visual/measure-parity.spec.ts [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+      results.push({
+        index,
+        measured,
+        painted,
+        delta: painted - measured,
+        text: text.slice(0, 60),
+        hasJoiner: lineEl.querySelector("[data-docx-joiner]") !== null,
+        runCount: lineEl.children.length,
+      });
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply tests/visual/measure-parity.spec.ts:135]: Fixed. `runCount` now counts the content spans the comparison already collected rather than `lineEl.children`. The direction matters: joiner spans and the list and break markers were each buying another half-pixel, so a repaired line was judged more loosely than the plain line beside it, which is backwards for the one line this spec exists to watch.

File: tests/visual/measure-parity.spec.ts [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    return results;
+  });
+}
+
+/** Load a document into the playground and wait for the first page to paint. */
+async function openFixture(page: Page, fixture: string): Promise<void> {
+  await page.goto(`/?file=${encodeURIComponent(fixture)}`);
+  await page.waitForSelector(".layout-page .layout-line", { timeout: 30_000 });
+  // Web fonts change advances, so a comparison taken mid-swap is meaningless.
+  await page.evaluate(() => document.fonts.ready);
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply tests/visual/measure-parity.spec.ts]: Fixed — the callback is now `async` and awaits without returning, so no `FontFaceSet` crosses the serialisation boundary.

File: .changeset/cursive-joining-face-change.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+"@stll/folio-core": patch
+---
+
+Keep cursive words joined when a run boundary splits them mid-word. Browsers shape across an inline box boundary only while no shaping-relevant property changes, so a bold, italic, resized or refaced run inside an Arabic word stopped shaping there and the word rendered as isolated letter forms; Word joins straight through the same boundary. The painter now emits a zero-width joiner on each side of such a boundary, each in its own span so run text nodes keep their exact ProseMirror offsets. Colour and underline boundaries (tracked changes, comment anchors) already shaped correctly and are left untouched. Joining classes are generated from the pinned Unicode Character Database rather than hand-listed, so combining marks are classified correctly.
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply .changeset/cursive-joining-face-change.md]: Reworded. Split into shorter sentences rather than only adding commas, which reads better in a changelog.
